src/_guides/language/effective-dart/design.md

```diff
@@ -952,13 +975,76 @@ class CallChainVisitor {
 }
 {% endprettify %}
 
+This guideline is deliberately softer than the previous one because the benefits
+of static typing are slightly reduced when code is confined within a single
+library. In practice, though, most Dart users type annotate all declarations,
+public or private.
+
+
+### AVOID annotating types for initialized local variables.
+
+Function bodies in modern code tend to be short, and the types of local
+variables are almost always obvious and inferrable from the initializing
+expression. Writing that type down is often simply visual noise. Instead, spend
+that energy choosing a great *name* for your local variable.
+
+{:.good-style}
+<?code-excerpt "misc/lib/effective_dart/design_good.dart (omit-types-on-locals)"?>
+{% prettify dart %}
+List<List<Ingredient>> possibleDesserts(Set<Ingredient> pantry) {
+  var desserts = <List<Ingredient>>[];
+  for (var recipe in cookbook) {
+    if (pantry.containsAll(recipe)) {
+      desserts.add(recipe);
+    }
+  }
+
+  return desserts;
+}
+{% endprettify %}
+
+{:.bad-style}
+<?code-excerpt "misc/lib/effective_dart/design_bad.dart (omit-types-on-locals)"?>
+{% prettify dart %}
+List<List<Ingredient>> possibleDesserts(Set<Ingredient> pantry) {
+  List<List<Ingredient>> recipes = <List<Ingredient>>[];
+  for (List<Ingredient> ingredients in cookbook) {
+    if (pantry.containsAll(ingredients)) {
+      recipes.add(ingredients);
+    }
+  }
+
+  return recipes;
+}
+{% endprettify %}
+
+If the local variable doesn't have an initializer, then its type can't be
+inferred. In that case, it *is* a good idea to annotate it. Otherwise, you get
+`dynamic` and lose the benefits of types.
+
+{:.good-style}
+<?code-excerpt "misc/lib/effective_dart/design_good.dart (uninitialized-local)"?>
+{% prettify dart %}
+List<AstNode> parameters;
+if (node is Constructor) {
+  parameters = node.signature;
+} else if (node is Method) {
+  parameters = node.parameters;
+}
+{% endprettify %}
+
 
-### AVOID annotating types on function expressions.
+### AVOID annotating inferrable parameter types on function expressions.
 
-The value of function expressions is their brevity. If a function is complex
-enough that types are needed to understand it, it should probably be a function
-statement or a method. Conversely, if it is short enough to be an expression, it
-likely doesn't need types.
+Anonymous functions are almost always immediately passed to a method expecting a
+function of some type. (If it isn't used immediately, it's usually worth making
+it a named function declaration.) When a function expression is created in a
+typed context, Dart tries to infer the function's parameter types based on the
+expected type.
+
+For example, when you pass a function expression to `Iterable.map()`, your
+function's parameter type is inferred based on the type of callback that
+`map()` expects and you don't need to annotate it:
```

Comment:
Maybe drop "and you don't need to annotate it"?

Reply:
Done.